src/context.ts

```diff
@@ -634,65 +596,33 @@ export class Context<
     arg?: StatusCode | ResponseOrInit,
     headers?: HeaderRecord
   ): Response {
-    // Optimized
-    if (this.#isFresh && !headers && !arg && this.#status === 200) {
-      return new Response(data, {
-        headers: this.#preparedHeaders,
-      })
-    }
-
-    if (arg && typeof arg !== 'number') {
-      const header = new Headers(arg.headers)
-      if (this.#headers) {
-        // If the header is set by c.header() and arg.headers, c.header() will be prioritized.
-        this.#headers.forEach((v, k) => {
-          if (k === 'set-cookie') {
-            header.append(k, v)
-          } else {
-            header.set(k, v)
-          }
-        })
-      }
-      const headers = setHeaders(header, this.#preparedHeaders)
-      return new Response(data, {
-        headers,
-        status: arg.status ?? this.#status,
-      })
-    }
-
-    const status = typeof arg === 'number' ? arg : this.#status
-    this.#preparedHeaders ??= {}
-
-    this.#headers ??= new Headers()
-    setHeaders(this.#headers, this.#preparedHeaders)
-
-    if (this.#res) {
-      this.#res.headers.forEach((v, k) => {
-        if (k === 'set-cookie') {
-          this.#headers?.append(k, v)
+    const responseHeaders = this.#res
+      ? new Headers(this.#res.headers)
+      : this.#preparedHeaders ?? new Headers()
+
+    if (typeof arg === 'object' && 'headers' in arg) {
+      const argHeaders = arg.headers instanceof Headers ? arg.headers : new Headers(arg.headers)
+      for (const [key, value] of argHeaders) {
+        if (key.toLowerCase() === 'set-cookie') {
+          responseHeaders.append(key, value)
         } else {
-          this.#headers?.set(k, v)
+          responseHeaders.set(key, value)
         }
-      })
-      setHeaders(this.#headers, this.#preparedHeaders)
+      }
     }
 
-    headers ??= {}
-    for (const [k, v] of Object.entries(headers)) {
-      if (typeof v === 'string') {
-        this.#headers.set(k, v)
-      } else {
-        this.#headers.delete(k)
-        for (const v2 of v) {
-          this.#headers.append(k, v2)
+    if (headers) {
+      for (const [key, value] of Object.entries(headers)) {
+        if (Array.isArray(value)) {
+          for (const v of value) responseHeaders.append(key, v)
```

Comment:
I think this is necessary to ensure compatibility.

```suggestion
          responseHeaders.delete(key)
          for (const v of value) {
            responseHeaders.append(key, v)
          }
```

Reply:
I updated the code following the former implementation and added the test: https://github.com/honojs/hono/pull/4100/commits/a722d539ff04126fd06b75f4a4f988efc03e8779